System/IO/BlockIO.hs

```diff
@@ -141,9 +143,8 @@ closeIOCtx IOCtx {ioctxURing, ioctxCloseSync} = do
         URing.closeURing uring
         putMVar ioctxURing Nothing
 
-data IOOp = IOOpRead  !Fd !FileOffset !(Ptr Word8) !ByteCount
-          | IOOpWrite !Fd !FileOffset !(Ptr Word8) !ByteCount
-  deriving Show
+data IOOp m = IOOpRead  !Fd !FileOffset !(MutableByteArray (PrimState m)) !Int !ByteCount
+            | IOOpWrite !Fd !FileOffset !(MutableByteArray (PrimState m)) !Int !ByteCount
```

Comment:
We need to document that these _must_ be pinned byte arrays, and indeed we should probably check that at runtime (it shouldn't be expensive).